### PR TITLE
Trying new approach for defining service clients

### PR DESCRIPTION
**Pull request by drtyh2o**
Ok, check this out. I think this perhaps resolves some of your usability concerns @wesleyfsmith @macsj200. This changes the way service interfaces are defined and clients are constructed.

Most notable changes are:

1) the service interface declaration is more concise and narrative:
```ts
export interface JurisdictionService {
  getJurisdiction: (jurisdiction_id: UUID) => JurisdictionDomainModel
  ...
}
```
Notice that there's no longer any references to any `Promise`, `ServiceResult` or the like.

2) when exporting the service client, rather than simply _redacting_ the extra service provider methods, you construct the client as follows:
``` ts
const { initialize, shutdown, ...service } = JurisdictionServiceProvider

export const JurisdictionServiceClient: ServiceClient<JurisdictionService> = {
  getJurisdiction: UnwrapServiceResult(service.getJurisdiction),
  ...
}
```

3) Finally, consuming the client methods no longer requires the additional function call to extract the result and service errors can be handled via _catch_ (with the benefit of type guarding if desired).
```ts
export const GetJurisdictionHandler = async (req: GetJurisdictionRequest, res: GetJurisdictionResponse) => {
  try {
    const { jurisdiction_id } = req.params
    const jurisdiction = await JurisdictionServiceClient.getJurisdiction(jurisdiction_id)
    const { version } = res.locals
    return res.status(200).send({ version, jurisdiction })
  } catch (error) {
    if (isServiceError(error)) {
      if (error.type === 'NotFoundError') {
        return res.status(404).send({ error })
      }
    }
    return res.status(500).send({ error })
  }
}
```

### CODE REVIEW

File: packages/mds-service-helpers/client/index.ts [hunk 1]
```diff
@@ -39,3 +40,11 @@ export const getServiceResult = async <R>(request: Promise<ServiceResponse<R>>):
...
   return response.result
 }
+
+// eslint-reason type inference requires any
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type AnyServiceResponseMethod = AnyFunction<Promise<ServiceResponse<any>>>
+
+export const UnwrapServiceResult = <M extends AnyServiceResponseMethod>(method: M) => async (
+  ...args: Parameters<M>
+): Promise<ReturnType<M> extends Promise<ServiceResponse<infer R>> ? R : unknown> => getServiceResult(method(...args))
```
Comment (twelch) [inline packages/mds-service-helpers/client/index.ts:50]: I've been staring at this for a while, deep for my brain.  I appreciate seeing your conversion of the jurisdiction client interface to see what it's doing.  What's your take on commenting these types of things?

Comment (drtyh2o) [reply packages/mds-service-helpers/client/index.ts:50]: Some comments probably wouldn’t hurt.

Comment (drtyh2o) [reply packages/mds-service-helpers/client/index.ts:50]: But basically it’s using type inference to convert a function that returns `Promise<ServiceResult<R>>` to a function with identical parameters that returns `Promise<R>`

Comment (twelch) [reply packages/mds-service-helpers/client/index.ts:50]: Thanks even that one-liner is helpful.  And points me to go learn about inference.  I'm often asking myself what kind of typing construct is this? To be able to go and learn more myself. The advanced TS typing docs feel vast still.

File: packages/mds-jurisdiction-service/client/index.ts [hunk 2]
```diff
@@ -14,9 +14,17 @@
...
 
-export const JurisdictionServiceClient: JurisdictionServiceInterface = client
+export const JurisdictionServiceClient: ServiceClient<JurisdictionService> = {
+  createJurisdiction: UnwrapServiceResult(service.createJurisdiction),
+  createJurisdictions: UnwrapServiceResult(service.createJurisdictions),
+  deleteJurisdiction: UnwrapServiceResult(service.deleteJurisdiction),
+  getJurisdiction: UnwrapServiceResult(service.getJurisdiction),
+  getJurisdictions: UnwrapServiceResult(service.getJurisdictions),
+  updateJurisdiction: UnwrapServiceResult(service.updateJurisdiction)
+}
```
Comment (avatarneil) [inline packages/mds-jurisdiction-service/client/index.ts:30]: Man, I would _love_ if there was a concise way to do this. Create a new object with the same methods as another, but applies a wrapper over them.

File: packages/mds-service-helpers/client/index.ts (hunk 1 above)
Comment (avatarneil) [reply packages/mds-service-helpers/client/index.ts:50]: @twelch I've found the typescript handbook to be fairly poorly structured. A lot of the super-cool advanced type usage (like `infer`) came around in [TS 2.8](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html) but while the additions are documented there, it's a bit tricky to find them when just scanning the docs.

File: packages/mds-jurisdiction-service/client/index.ts (hunk 2 above)
Comment (drtyh2o) [reply packages/mds-jurisdiction-service/client/index.ts:30]: Totally. I spent a bit of time trying, but opted to establish the boundary between service/consumer and then trying again in a later PR. Would love to land on something akin to:
```ts
asServiceClient<I>(provider: ServiceProvider<I>): ServiceClient<I>
```

**Pull request merged**